eval/src/errors.rs

```diff
@@ -0,0 +1,107 @@
+use crate::jml_type::JmlType;
+use miette::{Diagnostic, SourceSpan};
+use thiserror::Error;
+
+#[derive(Error, Diagnostic, Debug)]
+#[error("evaluation error")]
+pub enum EvalError {
+    #[diagnostic(transparent)]
+    TypeError(#[from] TypeError),
+
+    #[diagnostic(transparent)]
+    RuntimeError(#[from] RuntimeError),
+}
+
+#[derive(Error, Diagnostic, Debug)]
+#[error("type error")]
+
+pub struct TypeError {
+    #[label("Error occurred here")]
+    pub span: SourceSpan,
+
+    #[source]
+    #[diagnostic_source]
+    pub kind: TypeErrorKind,
+}
+
+#[derive(Debug, Error, Diagnostic, Clone)]
+pub enum TypeErrorKind {
+    #[error("Mismatched types: expected {expected:?}, found {found:?}")]
+    #[diagnostic(code(type_error::mismatched_types))]
+    MismatchedTypes {
+        expected: Vec<JmlType>,
+        found: JmlType,
+    },
+
+    #[error("Expected {expected_count} arguments, but got {actual_count}")]
+    #[diagnostic(
+        code(type_error::argument_count_mismatch),
+        help("Check the function call to ensure the correct number of arguments are provided.")
+    )]
+    ArgumentCountMismatch {
+        expected_count: usize,
+        actual_count: usize,
+    },
+
+    #[error("Type {found} is not ordered")]
+    #[diagnostic(code(type_error::not_ordered))]
+    NotOrderedType { found: JmlType },
```

Comment:
What are the "ordered" types?

Reply:
comparable, probably not my best choice for the naming, just try to follow Rust trait and Haskell Ord class